web-common/src/features/dashboards/url-state/convert-partial-explore-state-to-url-params.ts

```diff
@@ -0,0 +1,384 @@
+  partialExploreState: Partial<MetricsExplorerEntity>,
+  exploreSpec: V1ExploreSpec,
+  // We have quite a bit of logic in TimeControlState to validate selections and update them
+  // Eg: if a selected grain is not applicable for the current grain then we change it
+  // But it is only available in TimeControlState and not MetricsExplorerEntity
+  timeControlsState: TimeControlState | undefined,
+  // Used to remove params that are equal to defaults. This helps keep the url shorter.
+  defaultExploreUrlParams?: URLSearchParams,
+  // Used to decide whether to compress or not based on the full url length
+  urlForCompressionCheck?: URL,
+) {
+  const searchParams = new URLSearchParams();
+
+  maybeSetParam(
+    searchParams,
+    partialExploreState,
+    "activePage",
+    (ap) =>
+      ToURLParamViewMap[
+        FromActivePageMap[ap ?? DashboardState_ActivePage.DEFAULT]
+      ] ?? ExploreUrlWebView.Explore,
+  );
+
+  // timeControlsState will be undefined for dashboards without timeseries
+  if (timeControlsState?.selectedTimeRange) {
+    copyParamsToTarget(
+      toTimeRangesUrl(partialExploreState, timeControlsState),
+      searchParams,
+    );
+  }
+
+  if ("whereFilter" in partialExploreState) {
+    const expr = mergeDimensionAndMeasureFilters(
+      partialExploreState.whereFilter,
+      partialExploreState.dimensionThresholdFilters ?? [],
+    );
+    let filterParam = "";
+    if (expr && expr?.cond?.exprs?.length) {
+      filterParam = convertExpressionToFilterParam(
+        expr,
+        partialExploreState.dimensionsWithInlistFilter,
+      );
+    }
+
+    searchParams.set(ExploreStateURLParams.Filters, filterParam);
+  }
+
+  switch (partialExploreState.activePage) {
+    case DashboardState_ActivePage.UNSPECIFIED:
+    case DashboardState_ActivePage.DEFAULT:
+    case DashboardState_ActivePage.DIMENSION_TABLE:
+    case undefined:
+      copyParamsToTarget(
+        toExploreUrlParams(partialExploreState, exploreSpec),
+        searchParams,
+      );
+      break;
+
+    case DashboardState_ActivePage.TIME_DIMENSIONAL_DETAIL:
+      copyParamsToTarget(
+        toTimeDimensionUrlParams(partialExploreState),
+        searchParams,
+      );
+      break;
+
+    case DashboardState_ActivePage.PIVOT:
+      copyParamsToTarget(toPivotUrlParams(partialExploreState), searchParams);
+      // Since we do a shallow merge, we cannot remove time grain from the state for pivot as it is a deeper key.
+      // So this is a patch to remove it from the final url.
+      searchParams.delete(ExploreStateURLParams.TimeGrain);
+      break;
+  }
+
+  if (defaultExploreUrlParams) {
+    [...searchParams.entries()].forEach(([key, value]) => {
+      const defaultValue = defaultExploreUrlParams.get(key);
+      if (
+        (defaultValue === null && value !== "") ||
+        (defaultValue !== null && value !== defaultValue)
+      ) {
+        return;
+      }
+      searchParams.delete(key);
+    });
+  }
```

Comment:
Here, we should be reconciling two `Partial<ExploreState>` (our canonical data structure) rather than reconciling `URLSearchParams` (merely a user-facing encoding).